include/net/download_client.h

```diff
@@ -118,117 +163,76 @@ struct download_client_cfg {
  */
 struct download_client {
-	/** Protect shared variables. */
-	struct k_mutex mutex;
-
-	/** Socket descriptor. */
-	int fd;
-
-	/** Destination address storage */
-	struct sockaddr remote_addr;
-
-	/** Response buffer. */
-	char buf[CONFIG_DOWNLOAD_CLIENT_BUF_SIZE];
-	/** Buffer offset. */
-	size_t offset;
-
+	/** Client configuration options. */
+	struct download_client_cfg config;
+	/** Host configuration options. */
+	struct download_client_host_cfg host_config;
+	/** Host name, null-terminated. */
+	char hostname[CONFIG_DOWNLOAD_CLIENT_MAX_HOSTNAME_SIZE];
+	/** File name, null-terminated. */
+	char file[CONFIG_DOWNLOAD_CLIENT_MAX_FILENAME_SIZE];
 	/** Size of the file being downloaded, in bytes. */
 	size_t file_size;
 	/** Download progress, number of bytes downloaded. */
 	size_t progress;
+	/** Buffer offset. */
+	size_t buf_offset;
 
-	/** Server hosting the file, null-terminated.
-	 *  The host name must be kept in scope while download is going on.
-	 */
-	const char *host;
-	/** File name, null-terminated.
-	 *  The file name must be kept in scope while download is going on.
+	/** Download client transport, http, CoAP, MQTT, ...
+	 *  Store a pointer to the selected transport per DLC instance to avoid looking it up each
+	 *  call.
 	 */
-	const char *file;
-	/** Configuration options. */
-	struct download_client_cfg config;
-
-	/** Protocol for current download. */
-	int proto;
-
-	struct  {
-		/** Whether the HTTP header for
-		 * the current fragment has been processed.
-		 */
-		bool has_header;
-		/** The server has closed the connection. */
-		bool connection_close;
-		/** Is using ranged query. */
-		bool ranged;
-	} http;
-
-	struct {
-		/** CoAP block context. */
-		struct coap_block_context block_ctx;
+	struct dlc_transport *transport;
+	/** Transport parameters. */
+	uint8_t transport_internal[CONFIG_DOWNLOAD_CLIENT_TRANSPORT_PARAMS_SIZE];
 
-		/** CoAP pending object. */
-		struct coap_pending pending;
-	} coap;
-
-	/** Internal thread ID. */
-	k_tid_t tid;
+	/** Ensure that thread is ready for download */
+	struct k_sem event_sem;
+	/** Protect shared variables. */
+	struct k_mutex mutex;
+	/** Download client state. */
+	enum download_client_state state;
 	/** Internal download thread. */
 	struct k_thread thread;
-	/** Ensure that thread is ready for download */
-	struct k_sem wait_for_download;
+	/** Internal thread ID. */
+	k_tid_t tid;
 
 	/* Internal thread stack. */
-	K_THREAD_STACK_MEMBER(thread_stack,
-			      CONFIG_DOWNLOAD_CLIENT_STACK_SIZE);
-
-	/** Event handler. */
-	download_client_callback_t callback;
-
-	/** Set socket to native TLS */
-	bool set_native_tls;
-
-	/** Close the socket when finished. */
-	bool close_when_done;
-
-	enum {
-		DOWNLOAD_CLIENT_IDLE,
-		DOWNLOAD_CLIENT_CONNECTING,
-		DOWNLOAD_CLIENT_DOWNLOADING,
-		DOWNLOAD_CLIENT_FINISHED,
-		DOWNLOAD_CLIENT_CLOSING
-	} state;
+	K_THREAD_STACK_MEMBER(thread_stack, CONFIG_DOWNLOAD_CLIENT_STACK_SIZE);
```

Comment:
I think we should remove define for K_THREAD_STACK_MEMBER and use K_KERNEL_STACK_MEMBER directly where needed. K_THREAD_STACK_MEMBER was deprecated and removed from Zephyr so it's better we don't confuse people with that.
These stacks don't work in user mode according to Zephyr documentation but I guess we don't have kernel and user mode separation in nrf.

Reply:
Replaced `K_THREAD_STACK_MEMBER` with ` K_KERNEL_STACK_MEMBER`.